### ISSUE TITLE
an static asset will return 404 sometimes and 200 other times

### ISSUE THREAD

**Issue by rosenfeld** [puma/puma#1642]
I know this has been discussed already previously in issues #313 and #337, however, this time, I'd like to report an even more weird behavior.

We pre-compile our static assets with webpack before running the app with Puma and 2 workers (`-w 2 --preload`). We serve those assets using the Rack::Static middleware in a Roda application.

In the past few weeks I've noticed that sometimes the app wasn't loading because it was failing to download some JS and I had to re-deploy the app to fix it. Today I noticed that simply sending the USR1 signal to the app was enough to fix it, but I still don't understand what is causing this. I can currently reproduce the issue by issuing the same curl command multiple times and watching it return 200 sometimes and 404 other times for the same url. I suspect one of the workers is working as expected while the other is not but I have absolutely no idea on how to further debug what's happening.

If I try to modify the application, just after deploying it it's very unlikely I'll be able to reproduce it any time soon, so I'm wondering if there's something we could do to debug the issue without restarting the cluster. We use a blue-green deployment strategy so I don't have to restart this app until next deployment. Is there anything I can do to further track what could be causing those intermitent 404? I've made sure the issue has nothing to do with nginx because when I issue the requests inside the server itself to localhost:3000 I can see the exactly same behavior.

We're currently using Puma 3.7.1 in a Debian Linux Docker container also running in a Debian Linux host. We have been using this stack for over an year but just recently I've noticed those issues and this application is mostly on maintenance mode, so each deploy will basically introduce a few changes but there has been a while since the last time we upgraded our gems, for example. The server hasn't changed either.

Any suggestions would be very welcomed :)

Thanks in advance.

**Comment by alexdowad** [comment]
- What does the entire HTTP response, including all the HTTP headers, look like when you get a "spurious" 404?
- What does your entire middleware stack look like?
- The part of Rack which returns a 404 if a static file cannot be served looks like this (in `lib/rack/file.rb:46`):

```
      available = begin
        ::File.file?(path) && ::File.readable?(path)
      rescue SystemCallError
        false
      end

      if available
        serving(request, path)
      else
        fail(404, "File not found: #{path_info}")
      end
```

I would really like to know:

- Whether Rack is going into the `fail` branch
- And if so, what the return values of `File.file?` and `File.readable?` were
- Whether it went into the exception handler for `SystemCallError`

You can find out by editing the code in your gem cache and adding some logging print statements. Use `gem which rack/file.rb` to find where this file is kept on your server. After you successfully debug the problem, uninstall and reinstall the `rack` gem to wipe out your edits.

**Comment by rosenfeld** [comment]
Hi @alexdowad, thanks for the reply and sorry for not answering to that before but it has been a very atypical week around here with lots of urgent issues to work on lately. Although it's not finished yet I could find some time to respond to this issue.

First of all, unfortunately, no auto-reloading is set in production for performance issues, so it wouldn't work to update that file when the issue is present as any changes wouldn't take effect until I restart the app but restarting the app would make the issue go away anyway. So, I'm guessing your suggestion is to  perform some changes before starting the application, which can be a bit tricky since our deploy is automated, but I think I can make some changes to this file in the Docker container that is used to run the app. I'll try to figure out some way to let you know for sure that it's gone through this branch although I suspect it is because the message I was getting was exactly like that.

Since I no longer have a running application in that state, I'll have to wait for it to happen again before I can answer your questions.

With regards to the middleware stack it's a tricky question though. This is a Roda app, which makes use of the multi_run plugin:

http://roda.jeremyevans.net/rdoc/classes/Roda/RodaPlugins/MultiRun.html

There's a main app that is composed by several other Rack/Roda apps, each one with their own set of middlewares set up, including quite some custom middlewares specifically suited for our own app. So, depending on the route, a different set of middlewares would be in place. For the static assets route, this is served directly by the main app. The main app has the following middlewares included in this specific order:

- Bugsnag::Rack (https://github.com/bugsnag/bugsnag-ruby/blob/master/lib/bugsnag/integrations/rack.rb)
- ProtectSourceMaps (custom middleware, will include the source below)
- Rack::Static (see params below)
- Rack::Session::Cookie
The remaining ones below are custom ones, but I think they are irrelevant as they don't get run for the /assets path:
- DbSession
- RequestId
- RequestLogger

Here's the source for ProtectSourceMaps:

```ruby
# frozen-string-literal: true
# hide source-maps from non-authorized users

require 'rack'
require_relative 'db_session'

module Apps
  module Middleware
    # warning: when using a CDN, this should require some special handling for .map files, telling
    # the CDN that it shouldn't cache the request and always forward it to the server
    class ProtectSourceMaps
      def initialize(app)
        @app = app
      end

      def call(env)
        path = env['PATH_INFO']
        if (match = path.start_with?('/assets') && path.end_with?('.map'))
          return [ 403, {}, [] ] unless allow? env
        end
        status, headers, body = @app.call env
        if match
          headers = Rack::Utils::HeaderHash.new(headers)
          headers['cache-control'] = 'private, max-age=31536000'
        end
        [status, headers, body]
      end

      ALLOWED_IPS = [ '104.196.245.109', '104.196.254.247' ] # bugsnag servers:
      # https://docs.bugsnag.com/platforms/browsers/source-maps/
      def allow?(env)
        ip = env['HTTP_X_FORWARDED_FOR']&.split(',', 2)&.first
        return true if ALLOWED_IPS.include? ip
        session = DbSession::SessionHash.new Rack::Request.new env
        session[:roles]&.any?{|r| r['role'] == 'admin' }
      end
    end
  end
end
```

Here's how Rack::Static is used:

```ruby
use Rack::Static, urls: ['/assets', '/favicon.ico'], root: AppSettings.assets_path,
  header_rules: [
    [:all, {
      'cache-control' => 'public, max-age=31536000',
      'access-control-allow-origin' => '*',
    }]
  ]
```

So, maybe, for the time being, pehaps it would be better to simply copy the entire source of rack/file into my project and make the edits and use it instead of the one provided by the Rack gem. Then, once I have all details I can let you know and move back to the original implementation.

I'll try to find some time this week to perform those changes.

Thanks for the heads up.

**Comment by alexdowad** [comment]
@rosenfeld, from your comments it sounds like you have never been able to reproduce this problem in a local instance of the app, only in production. That is a bit of a pain.

Although your deploy is automated, can't you run `pumactl -p <PID> restart` on the server to restart your application instances? Then they will load the edited code.

You mentioned that you were able to reproduce the behavior by sending requests to your app with `curl`; I guess that does not work immediately after a restart? If so, you could write a little script to keep hitting your application with static file requests and somehow alert you as soon as it gets a 404 response back. Run that on your dev PC all day.

This is a tricky bug, but you can and will find and fix it. You just need a way to "shine a light" into the internals of your app and find out what is really happening.

**Comment by rosenfeld** [comment]
Hi @alexdowad, that's right, only in production and only a few times. Mind you that only about a dozen users seem to use the application in a given month and use it quite often. However, the requests to static resources are pretty rare since they're usually in those users' cache after the first visit after a new deployment, which probably explains why sometimes I notice and fix the issue even before our clients realize the problem.

Our monitoring scrips perform some actions in the app every minute and, to save some time in the script execution, we run a headless Chrome instance with cache enabled which means our monitoring scripts currently don't detect this issue as the static resources would be usually in cache, so I'm not alerted when that happens in production. I should probably either disable caching in the scripts or add a new check to specifically make sure I can fetch any static resource anyway.

Our app runs through a systemd service which basically runs Puma through a Docker container. When that issue happens, I noticed that running "systemctl reload app-name" would fix the issue and it basically runs "docker kill -s USR1 container-name", which I believe is what "pumactl restart" would do as well. That means it wouldn't help because as soon as I send the signal to Puma the issue would be fixed and I wouldn't be able to debug anything.

I can try adding a local check in my development machine to see if I can reproduce the issue, but I'm not that confident that I'll be able to :)

Ideally it would be great if I could somehow add remote breakpoints when adding some special params to the assets requests so that I could investigate what's happening when the issue happens. We use blue-green deployment, so the next time the issue arises I could perform a new deploy to the green version, for example, leaving the blue version as is so that I could debug it without interrupting the production services. The main problem is to understand how I can inspect the issue while the app is running without restarting the app, since restarting the app would immediately fix the issue, and I'm not talking about stopping and restarting the container, but by sending the USR1 signal to Puma.

So far I've instrumented the app to use a copy of 'rack/file' and 'rack/static' with some extra headers and logs to the original 'rack/file' so that it might give me a hint on why it's returning 404 sometimes. I intend to deploy this new version later today or tomorrow since it seems our client will be demonstrating our product today so I want to avoid any bugs caused by that change during the demonstration.

Also, I'm sorry but I sent you the middleware stack of another app, rather than the one experiencing this issue. For this one, there's only a single middleware in place before Rack::Static, which is a custom one:

```ruby
# frozen-string-literal: true

module Apps
  module Middleware
    class DefaultCharset
      def initialize(app)
        @app = app
      end

      def call(env)
        status, headers, body = result = @app.call(env)
        h, v = headers.find{|k, v| k =~ /\Acontent-type\z/i }
        if h && v !~ /charset/
          headers[h] = "#{v}; charset=UTF-8"
        end
        result
      end
    end
  end
end
```

Then Rack::Static was used like this:

```ruby
use Rack::Static, urls: ['/assets', '/favicon.ico'], root: 'public'
```

**Comment by rosenfeld** [comment]
I've just added this to my crontab, I'll let you know if I'll be ever notified about it:

```
* * * * * /bin/bash -c "curl -f -s http://127.0.0.1:3000/favicon.ico || notify-send 'static assets returning 404'"
```

**Comment by rosenfeld** [comment]
By the way, my crontab entry wasn't really working, I had to change it to set DISPLAY and XAUTHORITY env vars in order to actually get the notifications.

**Comment by rosenfeld** [comment]
I've added that to the instrumentation code, but I suspect I won't be able to confirm because, if I'm right, I won't experience this bug anymore now that I'm passing the full path to the `root` argument.

I'll try to find some time to submit a patch to Rack next week that will expand the path upon the initialization of the middleware.

I do have this Dir.chdir block in one place as I mentioned in the Rack issue, but it should get back to the original directory so, unless there's a Ruby bug preventing that from happen in some circumnstances, that wouldn't explain the bug. Let me further explain it. We use a blue-green deployment strategy. When the bug happened I've issued a new deployment to the green variation (assuming we were running the application in the blue variation). After I switch it to the green variation no more requests are dispatched to the blue variation, however, I still get the intermitent 404 responses when requesting those assets in the blue variation and I'm pretty sure no other requests are being sent to blue when green is the active one, so I'm pretty sure that if pwd had changed for one of the workers it remained changed, without rolling back to the original directory. I couldn't find any usage of Dir.chdir without the block being provided.

As I said in the other issue, my code is currently not thread-safe, and I intend to fix it next week, but I think Ruby core should make sure the original pwd is restored even if concurrent calls to `Dir.chdir{}` take place. Anyway, it's my mistake and I'll fix it next week hopefully.

Thanks :)
